<a href="https://colab.research.google.com/github/naveenkumark1/EVA4/blob/master/Week1/python101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
print("OM")

OM


This python notbook is for the python 101 practice code

This is a MUST Python Refresher. You must finish this notebook as a part of overall grading.

Majority of this notebook is inspired from: Stanford's Python Refresher which can be found at: https://cs231n.github.io/python-numpy-tutorial/

We have modified the original notebook and added some questions here. You need to find the solutions and provide answers in your Q1 - Python 101 Quiz.

# Python
Python is a high-level, dynamically typed multiparadigm programming language. Python code is often said to be almost like pseudocode, since it allows you to express very powerful ideas in very few lines of code while being very readable.
New Section

# Python Versions
There are currently two different supported versions of Python, 2.7 and 3.5. Somewhat confusingly, Python 3.0 introduced many backwards-incompatible changes to the language, so code written for 2.7 may not work under 3.5 and vice versa. For this class all code will use Python 3.5 or higher.

You can check your Python version at the command line by running.

In [0]:
!python --version

Python 3.6.9


# Basic data types

Like most languages, Python has a number of basic types including integers, floats, booleans, and strings. These data types behave in ways that are familiar from other programming languages.

Numbers: Integers and floats work as you would expect from other languages:

In [0]:
x = 37 # Initial Value of x
print(type(x)) # Prints "<class 'int'>"
print(x)       # Prints "3"
print(x + 1)   # Addition; prints "4"
print(x - 1)   # Subtraction; prints "2"
print(x * 2)   # Multiplication; prints "6" - Error
print(x ** 2)  # Exponentiation; prints "9" - Error
x += 1
print(x)  # Prints "4" - Error
x *= 2
print(x)  # Prints "8" - Error
y = 2.5
print(type(y)) # Prints "<class 'float'>"
print(y, y + 1, y * 2, y ** 2) # Prints "2.5 3.5 5.0 6.25"

x = x * y + x + y # Final value of x
print("Final Value of x is:", x)

<class 'int'>
37
38
36
74
1369
38
76
<class 'float'>
2.5 3.5 5.0 6.25
Final Value of x is: 268.5


# Quiz Question 1

With what integer value shall x be initialized as, such that Final Value of x is 268.5? Please mention the answer in INT value without decimal (e.g. 4, 2, 12, etc.).

Please answer this question in the Quiz Section.

# Booleans

Python implements all of the usual operators for Boolean logic, but uses English words rather than symbols (&&, ||, etc.):

In [0]:
t = True
f = False
print(type(t)) # Prints "<class 'bool'>"
print(t and f) # Logical AND; prints "False"
print(t or f)  # Logical OR; prints "True"
print(not t)   # Logical NOT; prints "False"
print(t != f)  # Logical XOR; prints "True"

<class 'bool'>
False
True
False
True


# Quiz Question 2

What is the value of

In [0]:
not ( not t + f * t)

True

# Strings

Python has great support for strings:

In [0]:
hello = 'hello'    # String literals can use single quotes
world = "world"    # or double quotes; it does not matter.
print(hello)       # Prints "hello"
print(len(hello))  # String length; prints "5"
hw = hello + ' ' + world  # String concatenation
print(hw)  # prints "hello world"
hw6 = '%s %s %d' % (hello, world, 6)  # sprintf style string formatting
print(hw6)  # prints "hello world 6"

location = 0
print(hw6[location])

location = -1
print(hw6[location])

location = 12
print(hw6[location])

location = len(hw6)-1
print(hw6[location])

hello
5
hello world
hello world 6
h
6
6
6


# Quiz Question 3

What should be the value of "location" such that print(hw6[location]) prints 6?

# Containers

Python includes several built-in container types: lists, dictionaries,sets, and tuples.

Lists A list is the Python equivalent of an array, but is resizeable and can contain elements of different types:

In [1]:
xs = [3, 1, 2]    # Create a list
print(xs, xs[2])  # Prints "[3, 1, 2] 2"
print(xs[-1])     # Negative indices count from the end of the list; prints "2"
xs[2] = 'foo'     # Lists can contain elements of different types
print(xs)         # Prints "[3, 1, 'foo']"
xs.append('bar')  # Add a new element to the end of the list
print(xs)         # Prints "[3, 1, 'foo', 'bar']"
x = xs.pop()      # Remove and return the last element of the list
print(x, xs)      # Prints "bar [3, 1, 'foo']"

[3, 1, 2] 2
2
[3, 1, 'foo']
[3, 1, 'foo', 'bar']
bar [3, 1, 'foo']


In [8]:
rng = 9
nums = list(range(rng))     # range is a built-in function that creates a list of integers
print(nums)               # Prints "[0, 1, 2, 3, 4]"
print(nums[2:4])          # Get a slice from index 2 to 4 (exclusive); prints "[2, 3]"
print(nums[2:])           # Get a slice from index 2 to the end; prints "[2, 3, 4]"
print(nums[:2])           # Get a slice from the start to index 2 (exclusive); prints "[0, 1]"
print(nums[:])            # Get a slice of the whole list; prints "[0, 1, 2, 3, 4]"
print(nums[:-1])          # Slice indices can be negative; prints "[0, 1, 2, 3]"
nums[2:4] = [8, 9]        # Assign a new sublist to a slice
print(nums)               # Prints "[0, 1, 8, 9, 4]"
print(sum(nums))

[0, 1, 2, 3, 4, 5, 6, 7, 8]
[2, 3]
[2, 3, 4, 5, 6, 7, 8]
[0, 1]
[0, 1, 2, 3, 4, 5, 6, 7, 8]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 8, 9, 4, 5, 6, 7, 8]
48


# Quiz Question 4

What should be the value of rng, such that print(sum(nums)) prints 48?

Loops

You can loop over elements of a list like this:

In [9]:
animals = ['cat', 'dog', 'monkey']
for animal in animals:
    print(animal)

cat
dog
monkey


If you want access to the index of each element within the body of a loop, use the built-in enumerate function:

In [10]:
animals = ['cat', 'dog', 'monkey']
for idx, animal in enumerate(animals):
    print('#%d: %s' % (idx + 1, animal))

#1: cat
#2: dog
#3: monkey


List comprehensions

When programming, frequently we want to transform one type of data into another. As a simple example, consider the following code that computes square numbers:

In [15]:
rng2 = 7
nums = range(rng2)
squares = []
for x in nums:
    squares.append(x ** 2)
print(sum(squares))

91
